app/routes/goal_routes.py

```diff
@@ -1 +1,142 @@
-from flask import Blueprint
\ No newline at end of file
+from flask import Blueprint, abort, make_response, request, Response
+from app.models.goal import Goal
+from app.models.task import Task
+from ..db import db
+from datetime import datetime
+
+goals_bp = Blueprint("goals_bp", __name__, url_prefix="/goals")
+
+
+@goals_bp.post("")
+def create_goal():
+    request_body = request.get_json()
+
+    try:
+        new_goal = Goal.from_dict(request_body)
+
+    except KeyError:
+        response = {"details": "Invalid data"}
+        abort(make_response(response, 400))
+
+    db.session.add(new_goal)
+    db.session.commit()
+
+    response = {"goal": new_goal.to_dict()}
+    return response, 201
+
+
+@goals_bp.get("")
+def get_all_goals():
+    query = db.select(Goal)
+
+    title_param = request.args.get("title")
+    sort = request.args.get("sort", "asc")  # Default to ascending order
+
+    query = db.select(Goal)
+
+    if title_param:
+        query = query.where(Goal.title.ilike(f"%{title_param}%"))
+
+    if sort == "desc":
+        query = query.order_by(Goal.title.asc())
+
+    goals = db.session.scalars(query)
+    goals_response = [goal.to_dict() for goal in goals]
+
+    return goals_response, 200
+
+
+@goals_bp.get("/<goal_id>")
+def get_single_goal(goal_id):
+    goal = validate_goal(goal_id)
+
+    return {"goal": goal.to_dict()}
+
+
+@goals_bp.put("/<goal_id>")
+def update_goal(goal_id):
+    goal = validate_goal(goal_id)
+    request_body = request.get_json()
+
+    goal.title = request_body["title"]
+
+    db.session.commit()
+    response = {"goal": goal.to_dict()}
+    return response, 200
+
+
+@goals_bp.patch("/<goal_id>/mark_complete")
+def mark_goal_complete(goal_id):
+    goal = validate_goal(goal_id)
+    goal.completed_at = datetime.now()
+    db.session.commit()
+
+    return {"goal": goal.to_dict()}, 200
+
+
+@goals_bp.patch("/<goal_id>/mark_incomplete")
+def mark_goal_incomplete(goal_id):
+    goal = validate_goal(goal_id)
+    goal.completed_at = None
+
+    db.session.commit()
+
+    return {"goal": goal.to_dict()}, 200
+
+
+@goals_bp.delete("/<goal_id>")
+def delete_goal(goal_id):
+    goal = validate_goal(goal_id)
+
+    db.session.delete(goal)
+    db.session.commit()
+    response = {"details": f'Goal {goal_id} "{goal.title}" successfully deleted'}
+    return response, 200
+
+
+@goals_bp.post("/<goal_id>/tasks")
+def assign_tasks_to_goal(goal_id):
+    goal = validate_goal(goal_id)
+    request_body = request.get_json()
+    task_ids = request_body.get("task_ids", [])
+
+    tasks = Task.query.filter(Task.id.in_(task_ids)).all()
+
+    goal.tasks.extend(tasks)
+    db.session.commit()
+
+    return {"id": goal.id, "task_ids": task_ids}, 200
```

Comment:
Notice that `task_ids` comes from the request body (line 101). Instead of taking the client-provided value, it would be nice to return a response that has task ids that we get from the goal that we fetched from the db on line 99. This way we know for sure that we're returning data from our source of truth (the DB) instead of just echoing back what the user sent to the API (what we get from `request_body.get("task_ids", [])`)

You'd have to grab tasks from goal like:
```py
tasks_from_goal = goal.tasks
task_ids_from_goal = [task.id for task in tasks_from_goal]

 return {
        "id": goal.id,
        "task_ids": task_ids_from_goal
}